### PR TITLE
Update NLP README to include information on API calls

### PR DESCRIPTION
**Pull request by magicznyleszek**
## Checklist

1. [ ] If you've added code that should be tested, add tests
2. [ ] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [ ] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a title and, if necessary, a description of your work suitable for publishing in our [release notes](https://community.kobotoolbox.org/tag/release-notes)
6. [ ] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)
7. [ ] Open an issue in the [docs](https://github.com/kobotoolbox/docs/issues/new) if there are UI/UX changes

## Description

Internal code documentation improvement

## Notes

See for context: https://chat.kobotoolbox.org/#narrow/stream/4-Kobo-Dev/topic/Front-end.20questions/near/456578

### CODE REVIEW

File: jsapp/js/components/processing/README.md [hunk 1]
```diff
@@ -85,3 +85,11 @@ translation.
...
    stored, if they don't, it would not.
+
+## Why we make so many calls?
+
+When we open NLP view (besides the stuff that happens for the first time; see "How it works (in a nutshell)" above), Front-end code is making multiple calls. Some of the calls are being made to the same endpoint, and it might look like a mistake (it's not). Here's the reasoning for each of them:
+
+1. `/api/v2/assets/<asset_uid>/data/?query=<query for rootuid>` - we need whole submission object for that single particular submission. This is the submission that is being displayed in the NLP view at this very moment. We use that data to display stuff (responses to all the other questions) in the sidebar.
+2. `/api/v2/assets/<asset_uid>/data/?fields=[<question_we_care_about>]` - we need responses to all rows that were activated in advanced features - and we need them for all submissions. This is needed for that dropdown in the top that allows users to select other question. It is also needed for the arrows in the top that allows users to change between submissions (we omit submissions without responses to currently selected question).
```
Comment (Akuukis) [inline jsapp/js/components/processing/README.md:94]: Do I understand right, that we call this multiple times for each question/row? Like,
- `/api/v2/assets/<asset_uid>/data/?fields=questionId1WithAdvancedFeatures`
- `/api/v2/assets/<asset_uid>/data/?fields=questionId2WithAdvancedFeatures`
- `/api/v2/assets/<asset_uid>/data/?fields=questionId3WithAdvancedFeatures`

And in contrast, the other two are called once?
- ...

Comment (magicznyleszek) [reply jsapp/js/components/processing/README.md]: I see now it should've been `questions_we_care_about` (plural) in here. I've improved the section a bit and added more info on calls being made in different scenarios :)

**Review by Akuukis — APPROVED** [review]
LGTM, works as described.

Repro:
- create a project with 3 audio question and 1 non-audio question
- submit 2 submissions
- open NLP view
- click through dropdown forth and back
- click arrows up and down
- watch devtools in the meantime

Notes:
- there's also `/api/v2/assets/<asset_id>/data/<id>/attachments/<id>`, IMHO it's ok to leave it not mentioned.

Out of scope for this PR, let's create new tasks for these?
- there is room for memoization, as clicking questions back and forth currently re-requests the query.
- there is room for optimization to skip `.../?fields=[<questions_we_care_about>]` query, because it seems to request only a subset of data already present at `.../?query=<query for rootuid>` response that could be reused instead.

**Comment by magicznyleszek** [comment]
@Akuukis Actually the room for skipping is with the two other calls. Let me explain.

## Explanation

I have a project with multiple audio questions and I have multiple submissions. When I open Single Processing through Data Table, FE is making three calls (as described above).

#### `/api/v2/assets/<asset_uid>/data/?query=<query for rootuid>`

The response I have:

```
{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "_id": 5,
            "formhub/uuid": "5c95654fc15a4df6869be4374b08ed7a",
            "start": "2024-09-25T10:54:22.258+02:00",
            "end": "2024-09-25T10:54:55.903+02:00",
            "A_small_note": "Test one",
            "First_audio": "test-spoken-10_54_42.m4a",
            "Second_audio": "test-characters-Pójdźże, kiń tę chmurność w głąb flaszy ™™ 1-10_54_48.mp3",
            "Third_audio": "ff-16b-2c-44100hz-10_54_54.aac",
            "__version__": "vghzcbNEewoHJ9h3M7DN5T",
            "meta/instanceID": "uuid:07b0db65-f7da-49b1-9a6b-5c413a8282f5",
            "_xform_id_string": "aL4VqojuLvPiB7yc7RphMW",
            "_uuid": "07b0db65-f7da-49b1-9a6b-5c413a8282f5",
            "_attachments": [
                {
                    "download_url": "http://kf.kobo.local/api/v2/assets/aL4VqojuLvPiB7yc7RphMW/data/5/attachments/1/",
                    "mimetype": "audio/x-m4a",
                    "filename": "kobo/attachments/5c95654fc15a4df6869be4374b08ed7a/07b0db65-f7da-49b1-9a6b-5c413a8282f5/test-spoken-10_54_42.m4a",
                    "instance": 5,
                    "xform": 3,
                    "id": 1,
                    "question_xpath": "First_audio"
                },
                {
                    "download_url": "http://kf.kobo.local/api/v2/assets/aL4VqojuLvPiB7yc7RphMW/data/5/attachments/3/",
                    "mimetype": "audio/aac",
                    "filename": "kobo/attachments/5c95654fc15a4df6869be4374b08ed7a/07b0db65-f7da-49b1-9a6b-5c413a8282f5/ff-16b-2c-44100hz-10_54_54.aac",
                    "instance": 5,
                    "xform": 3,
                    "id": 3,
                    "question_xpath": "Third_audio"
                }
            ],
            "_status": "submitted_via_web",
            "_geolocation": [
                null,
                null
            ],
            "_submission_time": "2024-09-25T08:54:56",
            "_tags": [],
            "_notes": [],
            "_validation_status": {},
            "_submitted_by": null,
            "_supplementalDetails": {
                "First_audio": {
                    "transcript": {
                        "value": "Test bangla",
                        "revisions": [],
                        "dateCreated": "2024-09-25 12:51:01",
                        "dateModified": "2024-09-25 12:51:01",
                        "languageCode": "bn"
                    }
                }
            }
        }
    ]
}
```

### `/advanced_submission_post/<asset_uid>/?submission=<submission_uid>`

The response I have:

```
{
    "First_audio": {
        "transcript": {
            "value": "Test bangla",
            "revisions": [],
            "dateCreated": "2024-09-25 12:51:01",
            "dateModified": "2024-09-25 12:51:01",
            "languageCode": "bn"
        }
    }
}
```

Note that this is exactly the same as `_supplementalDetails` from the call above. I already made a task for deduplicating that call: https://www.notion.so/kobotoolbox/Remove-duplicated-call-from-NLP-feature-10a7e515f65480eeaeb7c2a395bfe588?pvs=4

### `/api/v2/assets/<asset_uid>/data/?fields=[<questions_we_care_about>]`

The response I have:

```
{
    "count": 4,
    "next": null,
    "previous": null,
    "results": [
        {
            "_id": 8,
            "A_small_note": "Test four",
            "Second_audio": "test-spoken-10_55_25.m4a",
            "_uuid": "23d6536d-d721-4119-acff-5ee970711c66",
            "_supplementalDetails": {}
        },
        {
            "_id": 7,
            "A_small_note": "Test three",
            "First_audio": "test-song-8BP076-09-rushjet1-unknown_sector-10_55_16.mp3",
            "_uuid": "5b27c82d-c46f-4fde-a251-32cbebd60373",
            "_supplementalDetails": {}
        },
        {
            "_id": 6,
            "A_small_note": "Test two",
            "First_audio": "subset_01 - blocksize 4096-10_55_8.flac",
            "Second_audio": "test-10_55_3.mp3",
            "_uuid": "212d7779-3e97-478c-bb5b-451d2bc761ea",
            "_supplementalDetails": {}
        },
        {
            "_id": 5,
            "A_small_note": "Test one",
            "First_audio": "test-spoken-10_54_42.m4a",
            "Second_audio": "test-characters-Pójdźże, kiń tę chmurność w głąb flaszy ™™ 1-10_54_48.mp3",
            "Third_audio": "ff-16b-2c-44100hz-10_54_54.aac",
            "_uuid": "07b0db65-f7da-49b1-9a6b-5c413a8282f5",
            "_supplementalDetails": {
                "First_audio": {
                    "transcript": {
                        "value": "Test bangla",
                        "revisions": [],
                        "dateCreated": "2024-09-25 12:51:01",
                        "dateModified": "2024-09-25 12:51:01",
                        "languageCode": "bn"
                    }
                }
            }
        }
    ]
}
```

You can see that the last object is very similar to the first call, but it doesn't have `A_small_note` question response that we need.

**Comment by Akuukis** [comment]
Good point on skipping `advanced_submission_post` :ok_hand: 

> You can see that the last object is very similar to the first call, but it doesn't have `A_small_note` question response that we need.

I'm looking at `"A_small_note": "Test one",` in the first response in your comment and am confused :thinking: Did you miss it or is it somehow wrong?

**Comment by magicznyleszek** [comment]
The way to show it is to have a form with different kind of questions. Ideally with all the types ;)

**Comment by Akuukis** [comment]
Understood, ok I will trust you on that it's the case. So PR is LGTM, you've got a task for the other out of scope item :heavy_check_mark:

**Comment by magicznyleszek** [comment]
@Akuukis Just for completeness sake :D This is my minimal survey example (it has 2 questions: `audio` and `select_one`):

```
[
    {
        "type": "audio",
        "$kuid": "zd27t68",
        "label": [
            "Say something"
        ],
        "$qpath": "Say_something",
        "$xpath": "Say_something",
        "required": false,
        "$autoname": "Say_something"
    },
    {
        "type": "select_one",
        "$kuid": "bd4cm50",
        "label": [
            "What time is it?"
        ],
        "$qpath": "What_time_is_it",
        "$xpath": "What_time_is_it",
        "required": false,
        "$autoname": "What_time_is_it",
        "select_from_list_name": "ir7ga73"
    }
]
```

#### `/api/v2/assets/<asset_uid>/data/?query=<query for rootuid>`

```
{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "_id": 9,
            "formhub/uuid": "8a0244e844a94cddb3908723301c3ade",
            "start": "2024-09-25T16:49:45.752+02:00",
            "end": "2024-09-25T16:49:59.146+02:00",
            "Say_something": "test-16_49_57.mp3",
            "What_time_is_it": "it_s_12_00",
            "__version__": "vshJqiAqvEdLR4oSCZCytG",
            "meta/instanceID": "uuid:d292848f-b950-42b2-a1ce-fc55b504dc94",
            "_xform_id_string": "aKV6bp7AnVtUwXxrgpcfmB",
            "_uuid": "d292848f-b950-42b2-a1ce-fc55b504dc94",
            "_attachments": [
                {
                    "download_url": "http://kf.kobo.local/api/v2/assets/aKV6bp7AnVtUwXxrgpcfmB/data/9/attachments/8/",
                    "mimetype": "audio/mpeg",
                    "filename": "kobo/attachments/8a0244e844a94cddb3908723301c3ade/d292848f-b950-42b2-a1ce-fc55b504dc94/test-16_49_57.mp3",
                    "instance": 9,
                    "xform": 4,
                    "id": 8,
                    "question_xpath": "Say_something"
                }
            ],
            "_status": "submitted_via_web",
            "_geolocation": [
                null,
                null
            ],
            "_submission_time": "2024-09-25T14:50:05",
            "_tags": [],
            "_notes": [],
            "_validation_status": {},
            "_submitted_by": null,
            "_supplementalDetails": {
                "Say_something": {
                    "transcript": {
                        "value": "Jakiś polski test tutaj jest.",
                        "revisions": [],
                        "dateCreated": "2024-09-25 14:55:16",
                        "dateModified": "2024-09-25 14:55:16",
                        "languageCode": "pl"
                    }
                }
            }
        }
    ]
}
```

This call to get whole submission has `What_time_is_it` in it.

### `/advanced_submission_post/<asset_uid>/?submission=<submission_uid>`

```
{
    "Say_something": {
        "transcript": {
            "value": "Jakiś polski test tutaj jest.",
            "revisions": [],
            "dateCreated": "2024-09-25 14:55:16",
            "dateModified": "2024-09-25 14:55:16",
            "languageCode": "pl"
        }
    }
}
```

### `/api/v2/assets/<asset_uid>/data/?fields=[<questions_we_care_about>]`

```
{
    "count": 3,
    "next": null,
    "previous": null,
    "results": [
        {
            "_id": 11,
            "Say_something": "subset_01 - blocksize 4096-16_50_13.flac",
            "_uuid": "fcdb11ae-2710-464b-b2dc-c50ee3ac5a65",
            "_supplementalDetails": {}
        },
        {
            "_id": 10,
            "Say_something": "ff-16b-2c-44100hz-16_50_2.aac",
            "_uuid": "97df99be-2695-4298-a7df-70f7f7b82ee6",
            "_supplementalDetails": {}
        },
        {
            "_id": 9,
            "Say_something": "test-16_49_57.mp3",
            "_uuid": "d292848f-b950-42b2-a1ce-fc55b504dc94",
            "_supplementalDetails": {
                "Say_something": {
                    "transcript": {
                        "value": "Jakiś polski test tutaj jest.",
                        "revisions": [],
                        "dateCreated": "2024-09-25 14:55:16",
                        "dateModified": "2024-09-25 14:55:16",
                        "languageCode": "pl"
                    }
                }
            }
        }
    ]
}
```

This call to get some data for all submissions doesn't have `What_time_is_it`.

**Pull request merged**